### PR TITLE
Improve S2259: Variables of caught exceptions are known to be NotNull

### LINKED ISSUE #890: Rule S2259 FP: symbolic execution does not enter the `for` loop
### Description

The following code snippet seems to trigger a false positive S2259 `'lastEx' is null on at least one execution path.` on the last line with the throw statement:

```csharp
public FileStream Blah()
{
	IOException lastEx = null;
	for (int i = 0; i < 10; i++)
	{
		try
		{
			return File.Open("somepath", FileMode.Open);
		}
		catch (IOException e)
		{
			lastEx = e;
		}
	}

	throw new InvalidOperationException(lastEx.Message);
}
```
When the for loop is removed S2259 is not raised on the last line.

### Expected behavior

No S2259 raised on the last line as either the code has returned or the lastEx value has been set.

### Actual behavior

Rule S2259 is raised both using the scanner for MSBuild as well as using SonarLint.

### Related information

* SonarC# Version 6.5 (build 3766)
* Visual Studio Enterprise 2017 Version 15.4.1
* Scanner for MSBuild Version 3.0.2
* SonarLint 3.6.0.2584

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**
Partially fixes #890

Note: To really fix #890, we need to be able to detect, that the loop is always entered. We can fix it for a `do/while` loop now, for all other loops it is assumed that there is one bypass flow.

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationDispatcher.cs [hunk 1]
```diff
@@ -44,6 +44,7 @@ internal static class OperationDispatcher
         { OperationKindEx.ArrayElementReference, new ArrayElementReference() },
         { OperationKindEx.AnonymousObjectCreation, new Creation() },
         { OperationKindEx.Await, new Await() },
+        { OperationKindEx.CaughtException, new CaughtException() },
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationDispatcher.cs:47]: Let's add repro test cases to the rule as well.

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationDispatcher.cs:47]: I forgot to ask for the same thing in #6110, can you add one here?

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.TryCatchFinally.cs [hunk 2]
```diff
@@ -918,6 +918,26 @@ public void Finally_NestedLambda_ShouldNotFail()
...
+{
+    InstanceMethod();
+}
+catch(InvalidOperationException ex) when (Tag(""InFilter"", ex))
+{
+    Tag(""InCatch"", ex);
+}
+
+static bool Tag<T>(string name, T value) => true;
+";
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.TryCatchFinally.cs:935]: Nitpick
```suggestion
static bool Tag<T>(string name, T value) => true;";
```

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Loops.cs [hunk 3]
```diff
@@ -182,5 +182,67 @@ public void Loops_FalseConditionNotExecuted(string loop)
 Tag(""End"");";
             SETestContext.CreateCS(code).Validator.ValidateTagOrder("End");
         }
+
+        [TestMethod]
+        public void ForLoopWithTryCatchAndNullFlows()
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Loops.cs:187]: This change has nothing to do with loops. Loop was just a reproducer scenario.

Let's delete this and move this test case to the NRE rule test cases.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Loops.cs [hunk 4]
```diff
@@ -86,8 +86,8 @@ public void Loops_InstructionVisitedMaxTwice_EvenWithMultipleStates()
             var validator = SETestContext.CreateCS(code, ", int[] items", new AddConstraintOnInvocationCheck(), new PreserveTestCheck("condition"), new PreserveTestCheck("value")).Validator;
             validator.ValidateExitReachCount(4);
             var states = validator.TagStates("End");
-            var condition = states.SelectMany(x => x.SymbolsWith(BoolConstraint.False)).First();    // "False" is never set for "value"
-            var value = states.SelectMany(x => x.SymbolsWith(TestConstraint.First)).First();        // "First" is never set for "condition"
+            var condition = validator.Symbol("condition");
+            var value = validator.Symbol("value");
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Loops.cs:90]: Oh yeah

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/CaughtException.cs [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+namespace SonarAnalyzer
+{
+    internal sealed class CaughtException : SimpleProcessor<ICaughtExceptionOperationWrapper>
+    {
+        protected override ICaughtExceptionOperationWrapper Convert(IOperation operation) =>
+            ICaughtExceptionOperationWrapper.FromOperation(operation);
+
+        protected override ProgramState Process(SymbolicContext context, ICaughtExceptionOperationWrapper operation) =>
+            context.SetOperationConstraint(ObjectConstraint.NotNull);
+    }
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/CaughtException.cs:36]: Operation is not used => we can simplify
```suggestion
internal sealed class CaughtException : ISimpleProcessor
{
    public ProgramState Process(SymbolicContext context) =>
        context.SetOperationConstraint(ObjectConstraint.NotNull);
}
```

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/CaughtException.cs:36]: Looking at this, it's complete duplication with `Creation` class. With a little bit of imagination, caught exception operation creates the instance, so I'd remove this file and just reuse the Creation for this operation type too

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Let's delete some stuff

**Comment by pavel-mikula-sonarsource** [comment]
@martin-strecker-sonarsource Let's leave it partially-fixed and do not enter the loop topic. While migrating the CBDE, we'll build constraints for integers and that will ensure we will enter the loop at least once here.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/CaughtException.cs (hunk 5 above)
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/CaughtException.cs]: I wouldn't do that. Creation shouldn't set the "NotNull" constraint unconditionally (e.g. any value types or type parameters constrained to structs).

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationDispatcher.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationDispatcher.cs:47]: I dont' see a test case for the #6110

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.cs [hunk 6]
```diff
@@ -1343,6 +1358,27 @@ public int Go(string first, string second)
     }
 }
 
+public class Repo_890
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.cs:1361]:
```suggestion
// https://github.com/SonarSource/sonar-dotnet/issues/890
public class Repo_890
```

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM, please add a test case from the unresolved thread above before merging

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationDispatcher.cs (hunk 1 above)
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationDispatcher.cs:47]: Is this what you wanted? b47f979a870231f6cab6d3596ab3bd508fbb5b0f

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationDispatcher.cs:47]: Yes, exactly. Thanks

**Pull request merged**